### PR TITLE
Add an RP2350 HAL.

### PR DESCRIPTION
**Pull request by thejpster**
Taken from https://github.com/thejpster/rp-hal-rp2350-public, with minor tweaks.

### CODE REVIEW

**Comment by thejpster** [comment]
Hey, it finally completed all 98 jobs!

File: rp235x-hal/src/rom_data.rs [hunk 1]
```diff
@@ -0,0 +1,747 @@
+//! Functions and data from the RPI Bootrom.
+//!
+//! From the [rp235x datasheet](https://datasheets.raspberrypi.org/rp235x/rp235x-datasheet.pdf), Section 2.8.3.1:
+//!
+//! > The Bootrom contains a number of public functions that provide useful
+//! > rp235x functionality that might be needed in the absence of any other code
+//! > on the device, as well as highly optimized versions of certain key
+//! > functionality that would otherwise have to take up space in most user
+//! > binaries.
```
Comment (9names) [inline rp235x-hal/src/rom_data.rs:9]:
```suggestion
//! From the [rp235x datasheet](https://datasheets.raspberrypi.org/rp235x/rp235x-datasheet.pdf), Section 5.4:
//!
//! > Whilst some ROM space is dedicated to the implementation of the boot  
//! > sequence and USB/UART boot interfaces, the bootrom also contains public 
//! > functions that provide useful RP2350 functionality that may be useful for 
//! > any code or runtime running on the device.
```

File: rp235x-hal/src/rosc.rs [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+        self.device.ctrl().write(|w| w.enable().enable());
+
+        use fugit::RateExtU32;
+        self.transition(Enabled {
+            freq_hz: 6_500_000u32.Hz(),
+        })
+    }
+
+    /// Initializes the ROSC with a known frequency.
+    /// See sections 2.17.3. "Modifying the frequency", and 2.15.6.2. "Using the frequency counter"
```
Comment (9names) [inline rp235x-hal/src/rosc.rs:77]:
```suggestion
    /// See sections 8.3.4. "Modifying the frequency", and 8.3.8. "ROSC Counter"
```

File: rp235x-hal/src/spi.rs [hunk 3]
```diff
@@ -0,0 +1,579 @@
...
+        Self::MotorolaSpi(*f)
+    }
+}
+
+/// SPI frame format
+#[derive(Clone, Copy, PartialEq, Eq)]
+pub enum FrameFormat {
+    /// Motorola SPI format. See section 4.4.3.9 of rp235x datasheet.
+    MotorolaSpi(spi::Mode),
+    /// Texas Instruments synchronous serial frame format. See section 4.4.3.8 of rp235x datasheet.
```
Comment (9names) [inline rp235x-hal/src/spi.rs:78]:
```suggestion
    /// Texas Instruments synchronous serial frame format. See section 12.3.4.8 of rp235x datasheet.
```

File: rp235x-hal/src/spi.rs [hunk 4]
```diff
@@ -0,0 +1,579 @@
...
+    TexasInstrumentsSynchronousSerial,
+    /// National Semiconductor Microwire frame format. See section 4.4.3.14 of rp235x datasheet.
```
Comment (9names) [inline rp235x-hal/src/spi.rs:80]:
```suggestion
    /// National Semiconductor Microwire frame format. See section 12.3.4.14 of rp235x datasheet.
```

File: rp235x-hal/src/timer.rs [hunk 5]
```diff
@@ -0,0 +1,639 @@
+//! Timer Peripheral
+//!
+//! The Timer peripheral on rp235x consists of a 64-bit counter and 4 alarms.  
+//! The Counter is incremented once per microsecond. It obtains its clock source from the watchdog peripheral, you must enable the watchdog before using this peripheral.  
+//! Since it would take thousands of years for this counter to overflow you do not need to write logic for dealing with this if using get_counter.  
+//!
+//! Each of the 4 alarms can match on the lower 32 bits of Counter and trigger an interrupt.
+//!
+//! See [Chapter 4 Section 6](https://datasheets.raspberrypi.org/rp235x/rp235x_datasheet.pdf) of the datasheet for more details.
```
Comment (9names) [inline rp235x-hal/src/timer.rs:9]:
```suggestion
//! See [section 12.8](https://datasheets.raspberrypi.org/rp235x/rp235x_datasheet.pdf) of the datasheet for more details.
```

File: rp235x-hal/src/watchdog.rs [hunk 6]
```diff
@@ -0,0 +1,230 @@
+//! Watchdog
+//!
+//! The watchdog is a countdown timer that can restart parts of the chip if it reaches zero. This can be used to restart the
+//! processor if software gets stuck in an infinite loop. The programmer must periodically write a value to the watchdog to
+//! stop it from reaching zero.
+//!
+//! See [Chapter 4 Section 7](https://datasheets.raspberrypi.org/rp235x/rp235x_datasheet.pdf) of the datasheet for more details
```
Comment (9names) [inline rp235x-hal/src/watchdog.rs:7]:
```suggestion
//! See [Chapter 12 Section 9](https://datasheets.raspberrypi.org/rp235x/rp235x_datasheet.pdf) of the datasheet for more details
```

**Review by 9names — CHANGES_REQUESTED** [review]

File: rp235x-hal-examples/src/bin/adc_fifo_dma.rs [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+
+// UART related types
+use hal::uart::{DataBits, StopBits, UartConfig};
+
+/// Tell the Boot ROM about our application
+#[link_section = ".start_block"]
+#[used]
+pub static IMAGE_DEF: hal::block::ImageDef = hal::block::ImageDef::secure_exe();
+
+/// External high-speed crystal on the Raspberry Pi Pico board is 12 MHz. Adjust
```
[the remark this answers is not in the document]
Comment (thejpster) [reply rp235x-hal-examples/src/bin/adc_fifo_dma.rs]: No, let's call it what it is

**Comment by thejpster** [comment]
OK I think I hit them all. RIP all those GHA free minutes.

**Comment by jannic** [comment]
Some files got updates for rp2040 after you did the copy for rp235x.
Two examples I found were #831 for SPI and #798 for UART.
We could either update the pull request, or merge it first and then check for unintended divergences. What would you prefer?
(Maybe we can also do some more code deduplication?)

**Comment by 9names** [comment]
> We could either update the pull request, or merge it first and then check for unintended divergences.

Lets do it the latter way, it would be better as smaller update PRs.
Github really struggled when I was trying to review this mega PR, the sooner it's closed the better.

File: rp235x-hal/src/atomic_register_access.rs [hunk 8]
```diff
@@ -1,17 +1,17 @@
 //! Provide atomic access to peripheral registers
 //!
 //! This feature is not available for all peripherals.
-//! See [section 2.1.2 of the rp235x datasheet][section_2_1_2] for details.
+//! See [Section 2.1.3][section_2_1_3] of the RP2350 datasheet for details.
 //!
-//! [section_2_1_2]: https://datasheets.raspberrypi.com/rp235x/rp235x-datasheet.pdf#atomic-rwtype
+//! [section_2_1_3]: https://rptl.io/rp2350-datasheet#atomic-rwtype
```
Comment (9names) [inline rp235x-hal/src/atomic_register_access.rs:6]: Oh they have a short url now? Nice

File: rp235x-hal/src/lposc.rs [hunk 9]
```diff
@@ -1,7 +1,7 @@
 //! Low Power Oscillator (ROSC)
 //!
 //! See [Section
-//! 8.4](https://datasheets.raspberrypi.org/rp2350/rp2350_datasheet.pdf) for
+//! 8.4](https://datasheets.raspberrypi.org/.pdf) for
```
Comment (9names) [inline rp235x-hal/src/lposc.rs:4]:
```suggestion
//! 8.4](https://rptl.io/rp2350-datasheet) for
```

File: rp235x-hal/src/timer.rs [hunk 10]
```diff
@@ -6,7 +6,7 @@
 //!
 //! Each of the 4 alarms can match on the lower 32 bits of Counter and trigger an interrupt.
 //!
-//! See [Chapter 4 Section 6](https://datasheets.raspberrypi.org/rp235x/rp235x_datasheet.pdf) of the datasheet for more details.
+//! See [Section 12.8](https://datasheets.raspberrypi.org/.pdf) of the datasheet for more details.
```
Comment (9names) [inline rp235x-hal/src/timer.rs:9]:
```suggestion
//! See [Section 12.8](https://rptl.io/rp2350-datasheet) of the datasheet for more details.
```

**Review by 9names — APPROVED** [review]
LGTM, couple of copy-paste errors on datasheet URL but otherwise good to go

File: rp235x-hal/src/timer.rs (hunk 10 above)
Comment (thejpster) [reply rp235x-hal/src/timer.rs:9]: How was it even possible I did this? I'm baffled.

File: rp235x-hal/src/atomic_register_access.rs (hunk 8 above)
Comment (thejpster) [reply rp235x-hal/src/atomic_register_access.rs:6]: They always have had them. I saw it somewhere in the RP2040 HAL and thought "that's better".

**Comment by thejpster** [comment]
> Maybe we can also do some more code deduplication?

Yes, once this is in there's a *lot* of scope for de-duplication. I wasn't able to track the rp2040-hal whilst working in my private repo, so it sort of diverged. Then to open this PR I had to rebase a bunch of stuff and try and get closer to where you were (but you kept moving). I propose that after this goes in we do a `diff -r rp2040-hal rp235x-hal` and try and remove as many diffs as possible - either with changes to one of the HALs, or by moving things into rp-common-hal.

**Review by 9names — APPROVED** [review]

**Pull request merged**